File: cartservice/src/main/java/com/example/cartservice/application/services/CartService.java
package com.example.cartservice.application.services;

import com.example.cartservice.domain.models.Cart;
import com.example.cartservice.domain.models.CartItem;
import com.example.cartservice.domain.ports.persistence.CartRepository;
import com.example.commonlib.events.cart.CartCheckedOutEvent;
import com.example.commonlib.events.cart.CartItemDto;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.stereotype.Service;

import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
public class CartService {

    private final CartRepository cartRepository;
    private final KafkaTemplate<String, Object> kafkaTemplate;

    public Cart getCart(String userId) {
        return cartRepository.findById(userId)
                .orElseGet(() -> Cart.builder().id(userId).build());
    }

    public Cart addItemToCart(String userId, CartItem item) {
        Cart cart = getCart(userId);
        cart.addItem(item);
        return cartRepository.save(cart);
    }

    public Cart removeItemFromCart(String userId, Long productId) {
        Cart cart = getCart(userId);
        cart.removeItem(productId);
        return cartRepository.save(cart);
    }

    public Cart updateItemQuantity(String userId, Long productId, Integer quantity) {
        Cart cart = getCart(userId);
        cart.updateItemQuantity(productId, quantity);
        return cartRepository.save(cart);
    }

    public void clearCart(String userId) {
        cartRepository.deleteById(userId);
    }

    public Cart saveForLater(String userId, Long productId) {
        Cart cart = getCart(userId);
        cart.saveForLater(productId);
        return cartRepository.save(cart);
    }

    public Cart moveToCart(String userId, Long productId) {
        Cart cart = getCart(userId);
        cart.moveToCart(productId);
        return cartRepository.save(cart);
    }

    public void checkout(String userIdHeader, Long userId, String street, String city, String district, String country, String paymentMethod) {
        Cart cart = getCart(userIdHeader);
        if (cart.getItems() == null || cart.getItems().isEmpty()) {
            throw new RuntimeException("Cart is empty");
        }

        CartCheckedOutEvent event = CartCheckedOutEvent.builder()
                .userId(userId)
                .items(cart.getItems().stream()
                        .map(item -> CartItemDto.builder()
                                .productId(item.getProductId())
                                .productName(item.getProductName())
                                .sku(item.getSku())
                                .quantity(item.getQuantity())
                                .unitPrice(item.getUnitPrice())
                                .build())
                        .collect(Collectors.toList()))
                .shippingStreet(street)
                .shippingCity(city)
                .shippingDistrict(district)
                .shippingCountry(country)
                .paymentMethod(paymentMethod)
                .build();

        kafkaTemplate.send("cart.checked-out", userIdHeader, event);
        log.info("Published CartCheckedOutEvent for user {}", userId);

        clearCart(userIdHeader);
    }
}

File: frontend/features/cart/api/cartService.ts
import apiClient from '@/features/shared/api/client';
import { Cart, CartItem } from '@/features/shared/types';

interface AddCartItemRequest {
  productId: number;
  productName: string;
  sku?: string;
  quantity: number;
  unitPrice: number;
}

const emptyCart = (userId?: number | null): Cart => ({
  id: userId ? String(userId) : 'guest',
  items: [],
  savedItems: [],
  totalPrice: 0,
});

export const cartService = {
  getCart: async (): Promise<Cart> => {
    const response = await apiClient.get<Cart>('/api/cart');
    return {
      ...response.data,
      items: response.data.items ?? [],
      savedItems: response.data.savedItems ?? [],
      totalPrice: Number(response.data.totalPrice ?? 0),
    };
  },

  addItem: async (item: CartItem): Promise<Cart> => {
    const payload: AddCartItemRequest = {
      productId: Number(item.productId),
      productName: item.productName,
      sku: item.sku,
      quantity: item.quantity,
      unitPrice: item.unitPrice,
    };
    const response = await apiClient.post<Cart>('/api/cart/items', payload);
    return {
      ...response.data,
      items: response.data.items ?? [],
      savedItems: response.data.savedItems ?? [],
      totalPrice: Number(response.data.totalPrice ?? 0),
    };
  },

  updateItemQuantity: async (productId: string, quantity: number): Promise<Cart> => {
    const response = await apiClient.put<Cart>(`/api/cart/items/${productId}?quantity=${quantity}`);
    return {
      ...response.data,
      items: response.data.items ?? [],
      savedItems: response.data.savedItems ?? [],
      totalPrice: Number(response.data.totalPrice ?? 0),
    };
  },

  removeItem: async (productId: string): Promise<Cart> => {
    const response = await apiClient.delete<Cart>(`/api/cart/items/${productId}`);
    return {
      ...response.data,
      items: response.data.items ?? [],
      savedItems: response.data.savedItems ?? [],
      totalPrice: Number(response.data.totalPrice ?? 0),
    };
  },

  clearCart: async (): Promise<Cart> => {
    await apiClient.delete('/api/cart');
    return emptyCart();
  },

  saveForLater: async (productId: string): Promise<Cart> => {
    const response = await apiClient.post<Cart>(`/api/cart/items/${productId}/save-for-later`);
    return {
      ...response.data,
      items: response.data.items ?? [],
      savedItems: response.data.savedItems ?? [],
      totalPrice: Number(response.data.totalPrice ?? 0),
    };
  },

  moveToCart: async (productId: string): Promise<Cart> => {
    const response = await apiClient.post<Cart>(`/api/cart/items/${productId}/move-to-cart`);
    return {
      ...response.data,
      items: response.data.items ?? [],
      savedItems: response.data.savedItems ?? [],
      totalPrice: Number(response.data.totalPrice ?? 0),
    };
  },
};

export { emptyCart };
